core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -114,28 +123,45 @@ class HistoryServer(
    * this UI with the event logs in the provided base directory.
    */
   def initialize() {
-    attachPage(new HistoryPage(this))
+    if (!initialized.getAndSet(true)) {
```

Comment:
...Pulled it from the test suite, so the check is superflous, and cut it...

Looking at this more, I think the metrics setup should be moved to `bind()`, rather than `initialize()`, precisely because the metrics setup may be doing things in terms of thread startup, and you don't want that to happen so early in class construction. Doing that and the testing/review to make sure that the later-metrics-init won't break assumptions elsewhere.